src/update_agent/actor.rs

```diff
@@ -222,7 +222,16 @@ impl UpdateAgent {
         let can_fetch = self.strategy.can_check_and_fetch(&self.identity);
         let state_change = actix::fut::wrap_future::<_, Self>(can_fetch)
             .then(|can_fetch, actor, _ctx| actor.locked_upgrade(can_fetch, release))
-            .map(|res, actor, _ctx| res.map(|release| actor.state.update_staged(release)));
+            .map(|res, actor, _ctx| {
+                let success = res.is_ok();
+                if let Some(abandoned) = actor.state.deploy_attempted(success) {
```

Comment:
The API here feels awkward. We pass in whether the upgrade was a success or not and `deploy_attempted` not only changes the state, but also returns whether the update was abandoned?

I think it'd be easier to grok if we kept doing `update_staged()` here in the `is_ok()` case and otherwise pass it to a function called e.g. `actor.state.maybe_abandon()`? (And then log the warning in that function). WDYT? (To be clear, this is fine as is too!)

Reply:
I see that it feels a bit strange. I'm trying to keep the state-machine logic split from the actor logic, so that the former is easier to test/inspect on its own.

I reworked this code a bit so that the actor owns the result-branching and thus it is still in charge of calling either `state.update_stage()` or `state.record_failed_deploy()`, how does it look now?

Reply:
Yup, much better!